### PR TITLE
[Gui Editor] Adding Loading and Saving

### PR DESCRIPTION
**Pull request by msDestiny14**
Users can now save and load their gui locally or to the snippet server. This also be used in playgrounds. 
There is also code for the beginning of the parenting system.

### CODE REVIEW

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 1]
```diff
@@ -96,109 +98,76 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
...
+        let content = JSON.stringify(adt.serializeContent());
 
+        var xmlHttp = new XMLHttpRequest();
         xmlHttp.onreadystatechange = () => {
             if (xmlHttp.readyState == 4) {
                 if (xmlHttp.status == 200) {
                     var snippet = JSON.parse(xmlHttp.responseText);
-                    const oldId = material.snippetId;
-                    material.snippetId = snippet.id;
+                    const oldId = adt.snippetId || "_BLANK";
```
Comment (RaananW) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:121]: where is _BLANK coming from?

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 2]
```diff
@@ -96,109 +98,76 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
...
+                        navigator.clipboard.writeText(adt.snippetId);
                     }
 
                     let windowAsAny = window as any;
 
                     if (windowAsAny.Playground && oldId) {
                         windowAsAny.Playground.onRequestCodeChangeObservable.notifyObservers({
-                            regex: new RegExp(oldId, "g"),
-                            replace: material.snippetId
+                            regex: new RegExp(`parseFromSnippetAsync\\("${oldId}`, "g"),
```
Comment (RaananW) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:135]: can you explain the regex/replace paradigm?

I am asking since regex has `\\` (which in this case is actually a single `\` after a break), but the replace doesn't even have a single `\`

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 3]
```diff
@@ -301,6 +270,8 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
     }
 
     render() {
+
+        //var myScene=this.props.globalState.guiTexture.getScene();
```
Comment (RaananW) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:274]: nit - cleanup?

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 4]
```diff
@@ -344,15 +315,9 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
                         <ButtonLineComponent label="Save" onClick={() => {
                             this.save();
                         }} />
-                        {
-                            this.props.globalState.customSave &&
-                            <ButtonLineComponent label={this.props.globalState.customSave!.label} onClick={() => {
```
Comment (RaananW) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:349]: Was this added somewhere else? or is it simply not needed?

File: guiEditor/src/diagram/guiNode.ts [hunk 5]
```diff
@@ -102,19 +110,23 @@ export class GUINode {
     public constructor(globalState: GlobalState, public guiControl: Control) {
         this._globalState = globalState;
         this._ownerCanvas = this._globalState.workbench;
-        
+        this.x = guiControl.leftInPixels;
+        this.y = guiControl.topInPixels;
         guiControl.onPointerUpObservable.add(evt => {
             this.clicked = false;
             console.log("up");
```
Comment (RaananW) [inline guiEditor/src/diagram/guiNode.ts:115]: (not part of this PR, but...) is this console (and the others in this file)  needed?

File: guiEditor/src/diagram/guiNode.ts [hunk 6]
```diff
@@ -102,19 +110,23 @@ export class GUINode {
...
-        
+        this.x = guiControl.leftInPixels;
+        this.y = guiControl.topInPixels;
         guiControl.onPointerUpObservable.add(evt => {
             this.clicked = false;
             console.log("up");
         });
 
         guiControl.onPointerDownObservable.add( evt => {
+            if(!this._ownerCanvas.isUp) return;
```
Comment (RaananW) [inline guiEditor/src/diagram/guiNode.ts:119]: Our tslint/prettier should prevent this from passing lint, but we are not yet linting those files. Will you be able to run a formatter (prettier?) on these files?

File: guiEditor/src/diagram/guiNode.ts [hunk 7]
```diff
@@ -151,16 +165,53 @@ export class GUINode {
...
+        });
+
         return true;
         //evt.stopPropagation();
     }
 
+    renderContainer(): React.ReactNode {
+        if(!this._isContainer) return null;
+
+        return React.createElement(ContainerPropertyTabComponent, {
```
Comment (RaananW) [inline guiEditor/src/diagram/guiNode.ts:177]: any reason this isn't a tsx file and render jsx/html fragement here?

File: guiEditor/src/diagram/properties/containerPropertyComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import { IContainerComponentProps } from "./propertyComponentProps";
+
+export class ContainerPropertyTabComponent extends React.Component<IContainerComponentProps> {
+    constructor(props: IContainerComponentProps) {
+        super(props);
+    }
+
+    AddElement()
+    {
+        let newElementName = window.prompt("Please enter name of the element");
```
Comment (RaananW) [inline guiEditor/src/diagram/properties/containerPropertyComponent.tsx:16]: I assume this will be extended? Want to remove it from here for now?

File: guiEditor/src/guiEditor.ts [hunk 9]
```diff
@@ -57,7 +56,7 @@ export class GUIEditor {
 
         if (options.customLoadObservable) {
             options.customLoadObservable.add((data) => {
-                SerializationTools.Deserialize(data, globalState);
+                //TODO: Add deserilization here.
```
Comment (RaananW) [inline guiEditor/src/guiEditor.ts:59]: are you tracking this somehow?

File: guiEditor/src/stringTools.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    private static _SaveAs(blob: Blob, name: string, document: HTMLDocument) {
+        if ('download' in HTMLAnchorElement.prototype) {
+            var URL = window.URL || window.webkitURL;
+            var a = document.createElement('a');
+        
+            a.download = name;
+            a.rel = 'noopener'; // tabnabbing
+
+            a.href = URL.createObjectURL(blob)
+            setTimeout(() => { URL.revokeObjectURL(a.href) }, 4E4) // 40s
```
Comment (RaananW) [inline guiEditor/src/stringTools.ts:20]: 40 seconds?

File: guiEditor/src/stringTools.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
+export class StringTools {
```
Comment (RaananW) [inline guiEditor/src/stringTools.ts:1]: Don't we have some sort of file saver in the framework already? I am pretty sure I have seen similar code. Doesn't NME have the same thing?

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx (hunk 4 above)
Comment (msDestiny14) [reply guiEditor/src/components/propertyTab/propertyTabComponent.tsx:349]: Yes this is not needed at the moment but will come back eventaully.

File: guiEditor/src/diagram/guiNode.ts (hunk 5 above)
Comment (msDestiny14) [reply guiEditor/src/diagram/guiNode.ts:115]: For now I'd like to keep it in since I still do a lot of debugging around that. Is that considered proper?

File: guiEditor/src/diagram/guiNode.ts (hunk 7 above)
Comment (msDestiny14) [reply guiEditor/src/diagram/guiNode.ts:177]: Not really. Originally this was paired besides the other render but since that function is gone maybe I should rethink this.

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 12]
```diff
@@ -96,109 +98,76 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
...
+        let json = JSON.stringify(this.props.globalState.guiTexture.serializeContent());
+        StringTools.DownloadAsFile(this.props.globalState.hostDocument, json, "guiTexture.json");
     }
 
     saveToSnippetServer() {
-        /*const material = this.props.globalState.nodeMaterial;
-        const xmlHttp = new XMLHttpRequest();
-
-        let json = SerializationTools.Serialize(material, this.props.globalState);
+        var adt = this.props.globalState.guiTexture;
```
Comment (Popov72) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:113]: Use `let` or `const` instead of `var` (there are other occurrences of `var` througout the code).

Also, you should favor `const` over `let` whenever possible.

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx (hunk 1 above)
Comment (msDestiny14) [reply guiEditor/src/components/propertyTab/propertyTabComponent.tsx:121]: Ya. I'm not sure when there would be a time when snippetId doesn't exist.

File: guiEditor/src/guiEditor.ts (hunk 9 above)
Comment (msDestiny14) [reply guiEditor/src/guiEditor.ts:59]: You mean in github?

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx (hunk 2 above)
Comment (Popov72) [reply guiEditor/src/components/propertyTab/propertyTabComponent.tsx:135]: Actually the regex is `parseFromSnippetAsync\(...` but because it is hosted in a js string the `\` must be escaped to `\\`, so I think it's fine, it's really the `(` which is escaped in the expression (but something like `parseFromSnippetAsync\\s*\\(` would be safer).

But I don't understand why the previous code had to be changed: is it for additional safety (though there's little chance that the snipped id string would appear somewhere in the code without really being the snipped id we need to replace)?

File: guiEditor/src/diagram/guiNode.ts [hunk 13]
```diff
@@ -151,16 +163,53 @@ export class GUINode {
...
+        });
+    }
+
     public updateVisual()
     {
         this.guiControl.leftInPixels = this.x;
         this.guiControl.topInPixels = this.y;
     }
 
+    private isContainer() {
```
Comment (Popov72) [inline guiEditor/src/diagram/guiNode.ts:189]: Missing the leading underscore to the function name.

File: guiEditor/src/diagram/properties/containerPropertyComponent.tsx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import { LineContainerComponent } from "../../sharedUiComponents/lines/lineContainerComponent";
+import { TextLineComponent } from "../../sharedUiComponents/lines/textLineComponent";
+import { IContainerComponentProps } from "./propertyComponentProps";
+
+export class ContainerPropertyTabComponent extends React.Component<IContainerComponentProps> {
+    constructor(props: IContainerComponentProps) {
+        super(props);
+    }
+
+    AddElement()
```
Comment (Popov72) [inline guiEditor/src/diagram/properties/containerPropertyComponent.tsx:14]: Missing a public/protected/private modifier and probably the first letter should be lowercase (except if it is a static function?).

File: guiEditor/src/diagram/properties/propertyComponentProps.ts [hunk 15]
```diff
@@ -1,7 +1,13 @@
 import { Control } from "babylonjs-gui/2D/controls/control";
 import { GlobalState } from "../../globalState";
+import { GUINode } from "../guiNode";
 
 export interface IPropertyComponentProps {
     globalState: GlobalState;
-    guiControl: Control;
+    guiBlock: Control;
```
Comment (Popov72) [inline guiEditor/src/diagram/properties/propertyComponentProps.ts:7]: Did you have a specific reason to rename `guiControl` to `guiBlock`? `guiControl` seems better than `guiBlock` given that the type is `Control` . Or was it too generic?

**Comment by msDestiny14** [comment]
I removed some of the container related code since it's going to be reworked anyways in the future.

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx [hunk 16]
```diff
@@ -110,15 +109,15 @@ export class PropertyTabComponent extends React.Component<IPropertyTabComponentP
     }
 
     saveToSnippetServer() {
-        var adt = this.props.globalState.guiTexture;
+        let adt = this.props.globalState.guiTexture;
```
Comment (Popov72) [inline guiEditor/src/components/propertyTab/propertyTabComponent.tsx:112]: Most of the `let` you added could be `const` instead :)

If in doubt, just always use `const` and change to `let` if the transpiler does complain.

File: nodeEditor/src/diagram/display/inputDisplayManager.ts [hunk 17]
```diff
@@ -46,15 +45,19 @@ export class InputDisplayManager implements IDisplayManager {
...
 
         if (inputBlock.isAttribute) {
-            name = StringTools.GetBaseType(inputBlock.output.type);
+            name = this.GetBaseType(inputBlock.output.type);
         }
 
         return name;
     }
 
+    public GetBaseType(type: NodeMaterialBlockConnectionPointTypes): string {
```
Comment (Popov72) [inline nodeEditor/src/diagram/display/inputDisplayManager.ts:57]: You should either make this method static or have the first letter of its name be lowercase.

File: guiEditor/src/components/propertyTab/propertyTabComponent.tsx (hunk 2 above)
Comment (msDestiny14) [reply guiEditor/src/components/propertyTab/propertyTabComponent.tsx:134]: I think it was just implementations taken form different places since original the code was commented. I can revert back to the original (assuming it works)

Comment (msDestiny14) [reply guiEditor/src/components/propertyTab/propertyTabComponent.tsx:134]: Yep it does! :)

File: guiEditor/src/diagram/guiNode.ts (hunk 5 above)
Comment (RaananW) [reply guiEditor/src/diagram/guiNode.ts:111]: As long as it goes out for production, i am totally fine with it. Maybe just add a TODO to follow up

File: guiEditor/src/guiEditor.ts (hunk 9 above)
Comment (RaananW) [reply guiEditor/src/guiEditor.ts:59]: yep, some issue you can reference here, or some external todo list?
TODOs tend to stay in place...

File: guiEditor/src/diagram/guiNode.ts (hunk 5 above)
Comment (msDestiny14) [reply guiEditor/src/diagram/guiNode.ts:111]: Totally!

**Review by RaananW — APPROVED** [review]

File: guiEditor/src/guiEditor.ts (hunk 9 above)
Comment (msDestiny14) [reply guiEditor/src/guiEditor.ts:59]: Making git issues for next PR.

File: guiEditor/src/diagram/guiNode.ts (hunk 5 above)
Comment (msDestiny14) [reply guiEditor/src/diagram/guiNode.ts:111]: I'll make and issue for that.

**Comment by RaananW** [comment]
Just need to resolve the conflicts and hit the merge button :-)
